### PR TITLE
Issue #2475 Handle proxy specified without scheme

### LINKED ISSUE #2475: Protocol-less proxies are accepted by the command line parser but are not used
### General information

  * Minishift version: 1.19.0
  * OS: Linux
  * Hypervisor: VirtualBox

### Steps to reproduce

    HTTP_PROXY=1.1.1.1 ./minishift start --vm-driver virtualbox

### Expected

Should use the proxy and output

    -- Using proxy for the setup

### Actual

Doesn't use the proxy and doesn't output

    -- Using proxy for the setup

It works when you provide the proxy like 

    HTTP_PROXY=http://1.1.1.1 ./minishift start --vm-driver virtualbox

I can post a full log but I don't think it's necessary as this appears to be a parsing error of some kind. Same thing happens with --http-proxy and respective HTTPS versions.

I believe that 1.1.1.1 should be parsed as http://1.1.1.1 or alternatively throw an error. It should not silently fail, however.

### PR DESCRIPTION
**Pull request by gbraad**
Fixes #2475

Handle proxy specified without scheme

### CODE REVIEW

File: pkg/util/proxy.go [hunk 1]
```diff
@@ -170,13 +170,18 @@ func ValidateProxyURL(proxyUrl string) error {
 }
 
 // parseProxySpecialChar parse the URI and convert special char to hex
-func parseProxySpecialChar(uri string) string {
+func parseProxySpecialChar(uri string, scheme string) string {
```
Comment (praveenkumar) [inline pkg/util/proxy.go]: :+1:

File: pkg/util/proxy.go [hunk 2]
```diff
@@ -170,13 +170,18 @@ func ValidateProxyURL(proxyUrl string) error {
...
+	if uri == "" {
+		return ""
+	}
+
 	u, _ := url.Parse(uri)
-	scheme := fmt.Sprintf("%s://", u.Scheme)
 	if strings.HasPrefix(uri, scheme) {
+		scheme := fmt.Sprintf("%s://", u.Scheme)
```
Comment (praveenkumar) [inline pkg/util/proxy.go]: :+1:

**Comment by gbraad** [comment]
@svenstaro does this solve your issue?

**Comment by svenstaro** [comment]
Not quite happy with the fix: `--http-proxy=173.249.48.140` now works but `--http-proxy=173.249.48.140:8080` is an error. Either both of those work or both are errors.

File: pkg/util/proxy_test.go [hunk 3]
```diff
@@ -172,15 +172,16 @@ func Test_parse_special_character_uri(t *testing.T) {
...
-		{"", ""},
-		{"http://foo.com:3128", "http://foo.com:3128"},
-		{"http://user:F@oo!B#ar$@myserver:3128", "http://user:F@oo%21B%23ar$@myserver:3128"},
-		{"https://myuser:my#pass@foo.com:3128", "https://myuser:my%23pass@foo.com:3128"},
-		{"https://newuser:new(pas*)wrd@test.com:3128", "https://newuser:new%28pas%2A%29wrd@test.com:3128"},
+		{"", "", "http"},
+		{"http://foo.com:3128", "http://foo.com:3128", "http"},
+		{"http://user:F@oo!B#ar$@myserver:3128", "http://user:F@oo%21B%23ar$@myserver:3128", "http"},
+		{"https://myuser:my#pass@foo.com:3128", "https://myuser:my%23pass@foo.com:3128", "https"},
+		{"https://newuser:new(pas*)wrd@test.com:3128", "https://newuser:new%28pas%2A%29wrd@test.com:3128", "https"},
```
Comment (LalatenduMohanty) [inline pkg/util/proxy_test.go]: @gbraad shouldn't we test for  something like`{"foo.com:3128", "http://foo.com:3128", "http"}` ?

**Comment by gbraad** [comment]
this seems to be the standard behaviour for `!govalidator.IsURL(proxyUrl)`. We will have to add the prefix before the validation happens...

**Comment by gbraad** [comment]
@svenstaro I am not seeing that problem

```
minishift start --http-proxy=10.0.75.101:3128
-- Starting profile 'minishift'
-- Using proxy for the setup
-- Check if deprecated options are used ... OK
-- Checking if https://github.com is reachable ... OK
-- Checking if requested OpenShift version 'v3.9.0' is valid ... OK
-- Checking if requested OpenShift version 'v3.9.0' is supported ... OK
-- Checking if requested hypervisor 'kvm' is supported on this platform ... OK
-- Checking if KVM driver is installed ... 
   Driver is available at /usr/local/bin/docker-machine-driver-kvm ... 
   Checking driver binary is executable ... OK
-- Checking if Libvirt is installed ... OK
-- Checking if Libvirt default network is present ... OK
-- Checking if Libvirt default network is active ... OK
-- Checking the ISO URL ... OK
-- Checking if provided oc flags are supported ... OK
-- Starting the OpenShift cluster using 'kvm' hypervisor ...
-- Minishift VM will be configured with ...
   Memory:    1 GB
   vCPUs :    1
   Disk size: 5 GB
-- Starting Minishift VM ..............................-- Setting proxy information ... OK
 OK
-- Checking for IP address ... OK
-- Checking for nameservers ... OK
-- Checking if external host is reachable from the Minishift VM ... 
   Pinging 10.0.75.1 ... OK
-- Checking HTTP connectivity from the VM ... 
   Retrieving http://10.0.75.1:5000/v2/ ... FAIL
   VM cannot connect to external URL with HTTP
-- Checking if persistent storage volume is mounted ... OK
-- Checking available disk space ... 1% used OK
   Importing 'openshift/origin:v3.9.0' ....................... OK
   Importing 'openshift/origin-docker-registry:v3.9.0' ........ OK
   Importing 'openshift/origin-haproxy-router:v3.9.0' ........... OK
-- OpenShift cluster will be configured with ...
   Version: v3.9.0
-- Pulling the Openshift Container Image .. EXISTS
-- Copying oc binary from the OpenShift container image to VM ...... OK
-- Starting OpenShift cluster ................
```

**Comment by gbraad** [comment]
@svenstaro settings take precedence, so check what is set with `m config view`

File: pkg/util/proxy_test.go (hunk 3 above)
Comment (gbraad) [reply pkg/util/proxy_test.go]: Added, but this problem does not occur here. Let's see what CI says...

**Pull request merged**